File: InsertionSortList.java
/**
 * Created by yhf on 15/5/19.
 */

/*
Sort a linked list using insertion sort.
*/

public class InsertionSortList {
    public static ListNode insertionSortList(ListNode head) {
        ListNode newHead = new ListNode(Integer.MIN_VALUE);
        ListNode cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            ListNode p = newHead;
            while (p.next != null && p.next.val < cur.val) {
                p = p.next;
            }
            ListNode temp = p.next;
            p.next = cur;
            cur.next = temp;
            cur = next;
        }
        return newHead.next;
    }

    public static void main(String[] args) {
        ListNode n4 = new ListNode(4);
        ListNode n3 = new ListNode(3);
        ListNode n2 = new ListNode(2);
        ListNode n1 = new ListNode(1);
        n4.next = n3;
        n3.next = n2;
        n2.next = n1;
        ListNode head = insertionSortList(n4);
    }
}

File: InsertionSortList.py
# encoding: utf-8
"""
Sort a linked list using insertion sort.
"""
from ListNode import ListNode


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def insertionSortList(self, head):
        # Time Limit Exceeded
        dummy_head = ListNode(0)
        cur = head
        while cur:
            next = cur.next
            p = dummy_head
            while p.next and p.next.val < cur.val:
                p = p.next
            temp = p.next
            p.next = cur
            cur.next = temp
            cur = next
        return dummy_head.next

    def insertionSortList2(self, head):
        if not head:
            return head
        dummy_head = ListNode(0)
        dummy_head.next = head
        cur = head
        while cur.next:
            if cur.next.val >= cur.val:  # 如果链表升序，cur就向后移动
                cur = cur.next
            else:
                p = dummy_head
                while p.next.val < cur.next.val:
                    p = p.next
                temp = cur.next
                cur.next = temp.next
                temp.next = p.next
                p.next = temp
        return dummy_head.next


if __name__ == '__main__':
    n1 = ListNode(10, 9, 8, 7, 6, 5, 4)
    print Solution().insertionSortList2(n1)
